src/framework/mlt_image.c

```diff
@@ -0,0 +1,374 @@
+#include <string.h>
+
+/** Allocate a new Image object.
+ *
+ * \return a new image object with default values set
+ */
+
+mlt_image mlt_image_new()
+{
+	mlt_image self = calloc( 1, sizeof(struct mlt_image_s) );
+	self->close = free;
+	return self;
+}
+
+/** Destroy an image object created by mlt_image_new().
+ *
+ * \public \memberof mlt_image_s
+ * \param self the Image object
+ */
+
+void mlt_image_close( mlt_image self )
+{
+	if ( self)
+	{
+		if ( self->release_data )
+		{
+			self->release_data( self->data );
+		}
+		if ( self->close )
+		{
+			self->close( self );
+		}
+	}
+}
+
+/** Set the most common values for the image.
+ *
+ * Less common values will be set to reasonable defaults.
+ *
+ * \public \memberof mlt_image_s
+ * \param self the Image object
+ * \param data the buffer that contains the image data
+ * \param format the image format
+ * \param width the width of the image
+ * \param height the height of the image
+ */
+
+void mlt_image_set_values( mlt_image self, void* data, mlt_image_format format, int width, int height )
+{
+	self->data = data;
+	self->format = format;
+	self->width = width;
+	self->height = height;
+	self->release_data = NULL;
+	self->close = NULL;
+	mlt_image_format_planes( self->format, self->width, self->height, self->data, self->planes, self->strides );
+}
+
+/** Get the most common values for the image.
+ *
+ * \public \memberof mlt_image_s
+ * \param self the Image object
+ * \param[out] data the buffer that contains the image data
+ * \param[out] format the image format
+ * \param[out] width the width of the image
+ * \param[out] height the height of the image
+ */
+
+void mlt_image_get_values( mlt_image self, void** data, mlt_image_format* format, int* width, int* height )
+{
+	*data = self->data;
+	*format = self->format;
+	*width = self->width;
+	*height = self->height;
+}
+
+/** Allocate the data field based on the other properties of the Image.
+ *
+ * If the data field is already set, and a destructor function exists, the data
+ * will be released. Else, the data pointer will be overwritten without being
+ * released.
+ *
+ * After this function call, the release_data field will be set and can be used
+ * to release the data when necessary.
+ *
+ * \public \memberof mlt_image_s
+ * \param self the Image object
+ */
+
+void mlt_image_alloc_data( mlt_image self )
+{
+	if ( !self ) return;
+
+	if ( self->release_data )
+	{
+		self->release_data( self->data );
+	}
+
+	int size = mlt_image_calculate_size( self );
+	self->data = mlt_pool_alloc( size );
+	self->release_data = mlt_pool_release;
+	mlt_image_format_planes( self->format, self->width, self->height, self->data, self->planes, self->strides );
+}
+
+/** Calculate the number of bytes needed for the Image data.
+ *
+ * \public \memberof mlt_image_s
+ * \param self the Image object
+ * \return the number of bytes
+ */
+
+int mlt_image_calculate_size( mlt_image self )
+{
+	int height = self->height + 1; // Legacy bug workaround
```

Comment:
A couple of weeks ago, I added the removal of this `+ 1` to the v7 list. It causes too many bugs on code contributions and changes (most recent with RGBA in qimage). This is not the only occurrence.

Reply:
Yes! I am eager to do that, too. But how do we know what bugs this avoided so that we can fix them? Do we just remove the +1 and wait for bugs to present themselves?

Reply:
I did a git log message search on "height" and I did not find the reason for it, but I did find several crash fixes to workaround its unexpected result. I also searched my gmail, and I only found my vague answer to someone's question about it:

> Some image processing operations that are data-parallel (multiple byte) may overshoot the logical end of the image buffer.

It might be my old scaler in the gdk module or some of the things that use "Duff's device." I do not recall because it goes back so far! It seems to cause more harm than good, and there might be regressions, but we need to make this change some time, and it seems better to do it now than in a supposed stable branch.

I finally traced it back to my commit from 2004
https://github.com/mltframework/mlt/commit/74268cd86c30a0af9f03981c799aa4bba253c6e9

In this early development we made rather large commits without a lot of details in commit messages or comments while just cranking out the code. Now, I recall there was something related to field order conversion that involved shifting the image or something. Check this out, in particular "// Offset the image pointer by one line":

https://github.com/mltframework/mlt/commit/74268cd86c30a0af9f03981c799aa4bba253c6e9#diff-4ef9f584ebce981fe664a34a8222b8c6464718df6ed21aa2acce0a7ce5c6b0e7L41

At this time we were also trying to optimize things to meet our goals for the customer with a tight deadline and no certainty we would be doing anything further with MLT afterwards. This code is long since removed and not how we do field order correction today.

Reply:
I support this. I added it to the task list: https://github.com/mltframework/mlt/issues/655
I agree that we may just have to wait for the regressions to be presented to us. But 7.0 is the time to break it.